framework/src/base/FEProblemBase.C

```diff
@@ -4401,6 +4400,10 @@ FEProblemBase::possiblyRebuildGeomSearchPatches()
     {
       case 0: // Never
         break;
+      case 3: // Nonlinear iteration
+        // This case is included in PenetrationLocator.C as the subset of slave nodes
+        // for which penetration is not detected at the current iteration is required.
```

Comment:
Is there any benefit to falling into case 1 (always) to update the patch here, rather than rechecking the patches for the nodes that aren't in the patch later on?

Reply:
Case 1 (always) option deletes the existing nearest node locators and starts a new contact patch search for all the slave nodes at the start of every timestep. This could be more computational expensive than just updating the patch for the subset of slave nodes that move outside the contact patch. 

Also this happens only at the start of the time step. So if the node moves outside the contact patch during the nonlinear iterations within a time step, this option does not help with that.

Reply:
Maybe we should use the `auto` option then. I think the main benefit would be that we would start out with a reasonable guess of the off-diagonal Jacobian entries.

Reply:
I added `reinitBecauseOfGhostingOrNewGeomObjects(); ` and `initPetscOutput();` and this seems to take care of augmenting the Jacobian sparsity without messing up with the ghosted elements.

The slave nodes - neighbor nodes map from the end of the previous time step is used to augment the Jacobian sparsity at the start of the time step.